client/client_test.go

```diff
@@ -1354,6 +1354,352 @@ func TestSubscribePredefined(t *testing.T) {
 	}
 }
 
+func TestUnsubscribe(t *testing.T) {
+	assert := assert.New(t)
+
+	clientID := "test-client"
+	topic := "test/a"
+	qos := uint8(0)
+
+	stp := newTestSetup(t, clientID)
+	defer stp.cancel()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		stp.connect(clientID)
+
+		// client --SUBSCRIBE--> GW
+		subscribe := stp.recv().(*msgs.SubscribeMessage)
+		subscribe.QOS = qos
+		assert.Equal([]byte(topic), subscribe.TopicName)
+		assert.Equal(msgs.TIT_STRING, subscribe.TopicIDType)
+
+		// client <--SUBACK-- GW
+		suback := msgs.NewSubackMessage(1, 0, msgs.RC_ACCEPTED)
+		suback.CopyMessageID(subscribe)
+		stp.send(suback)
+
+		// client <--PUBLISH-- GW
+		msgID := uint16(123)
+		publish := msgs.NewPublishMessage(suback.TopicID,
+			msgs.TIT_REGISTERED, []byte(""), qos, false, false)
+		publish.SetMessageID(msgID)
+		stp.send(publish)
```

Comment:
This block is not needed IMHO. I think, it's enough to simulate subscribe and then unsubscribe.

If you agree, you can also remove everything related to `callbackFired` below...

Reply:
Done.